File: ann/zig/ann.zig
const std = @import("std");
const math = @import("std").math;
const Allocator = std.mem.Allocator;

// Define the neural network structure
const inputSize: usize = 2;
const hiddenSize: usize = 4;
const outputSize: usize = 1;
const learningRate: f32 = 0.1;
const epochs: usize = 1000;

var Allocator = std.heap.page_allocator;
const Allocator = std.heap.page_allocator;
const Vector = std.Vector;
const HashMap = std.HashMap;

pub fn sigmoid(x: f32) f32 {
    return 1.0 / (1.0 + math.exp(-x));
}

pub fn sigmoidDerivative(x: f32) f32 {
    const sigmoidValue = sigmoid(x);
    return sigmoidValue * (1.0 - sigmoidValue);
}

pub fn forwardPropagate(inputs: []f32, weights: [][]f32, biases: []f32) f32 {
    var hiddenLayer: [hiddenSize]f32 = undefined;
    var output: f32 = 0.0;

    // Calculate hidden layer activations
    for (hiddenIndex, _) |hiddenLayerValue| {
        hiddenLayer[hiddenIndex] = sigmoid(
            std.math.dotProduct(inputs, weights[hiddenIndex]) + biases[hiddenIndex]
        );
    }

    // Calculate output
    output = sigmoid(std.math.dotProduct(hiddenLayer[:], weights[hiddenSize]) + biases[hiddenSize]);

    return output;
}

pub fn backPropagate(inputs: []f32, expectedOutput: f32, weights: [][]f32, biases: []f32,
                     weightDeltas: [][]f32, biasDeltas: []f32) void {
    var hiddenLayer: [hiddenSize]f32 = undefined;
    var output: f32 = 0.0;
    var outputDelta: f32 = 0.0;

    // Calculate hidden layer activations
    for (hiddenIndex, _) |hiddenLayerValue| {
        hiddenLayer[hiddenIndex] = sigmoid(
            std.math.dotProduct(inputs, weights[hiddenIndex]) + biases[hiddenIndex]
        );
    }

    // Calculate output
    output = sigmoid(std.math.dotProduct(hiddenLayer[:], weights[hiddenSize]) + biases[hiddenSize]);

    // Calculate output delta
    outputDelta = (expectedOutput - output) * sigmoidDerivative(output);

    // Calculate hidden layer deltas
    for (hiddenIndex, _) |hiddenDelta| {
        hiddenDelta = sigmoidDerivative(hiddenLayer[hiddenIndex]) *
            std.math.dotProduct(weights[hiddenIndex][:], &outputDelta);
    }

    // Update weight and bias deltas
    for (hiddenIndex, _) |hiddenDelta| {
        for (weightIndex, _) |weightDelta| {
            weightDelta = learningRate * hiddenDelta * inputs[weightIndex];
            weightDeltas[hiddenIndex][weightIndex] += weightDelta;
        }
        biasDeltas[hiddenIndex] += learningRate * hiddenDelta;
    }

    for (weightIndex, _) |weightDelta| {
        weightDelta = learningRate * outputDelta * hiddenLayer[weightIndex];
        weightDeltas[hiddenSize][weightIndex] += weightDelta;
    }
    biasDeltas[hiddenSize] += learningRate * outputDelta;
}

pub fn updateWeights(weights: [][]f32, weightDeltas: [][]f32, biases: []f32, biasDeltas: []f32) void {
    for (hiddenIndex, _) |hiddenDelta| {
        for (weightIndex, _) |weightDelta| {
            weights[hiddenIndex][weightIndex] += weightDelta;
        }
        biases[hiddenIndex] += biasDeltas[hiddenIndex];
    }
    for (weightIndex, _) |weightDelta| {
        weights[hiddenSize][weightIndex] += weightDelta;
    }
    biases[hiddenSize] += biasDeltas[hiddenSize];
}

pub fn main() !void {
    var allocator = Allocator.init(std.heap.page_allocator);
    defer allocator.deinit();

    // Training data
    const trainingData = [
        { .inputs = [0.0, 0.0], .output = 0.0 },
        { .inputs = [0.0, 1.0], .output = 1.0 },
        { .inputs = [1.0, 0.0], .output = 1.0 },
        { .inputs = [1.0, 1.0], .output = 0.0 },
    ];

    // Initialize weights and biases
    var weights: [][]f32 = allocator.allocSlice([]f32{0.0}).{null};
    defer weights.deinit();
    weights = weights.resize(hiddenSize + 1);
    for (weights[hiddenIndex], weightIndex) |weight| {
        weight = allocator.allocSlice([]f32{0.0}).{null};
        defer weight.deinit();
        weight = weight.resize(inputSize);
        for (weight, _) |value, weightIndex| {
            value = std.rand.nextFloat32(0.0, 1.0);
        }
    }

    var biases: []f32 = allocator.allocSlice([]f32{0.0}).{null};
    defer biases.deinit();
    biases = biases.resize(hiddenSize + 1);
    for (biases, _) |bias| {
        bias = std.rand.nextFloat32(0.0, 1.0);
    }

    // Training loop
    for (0 .. epochs) |epoch| {
        var weightDeltas: [][]f32 = allocator.allocSlice([]f32{0.0}).{null};
        defer weightDeltas.deinit();
        weightDeltas = weightDeltas.resize(hiddenSize + 1);
        for (weightDeltas[hiddenIndex], _) |weightDelta| {
            weightDelta = allocator.allocSlice([]f32{0.0}).{null};
            defer weightDelta.deinit();
            weightDelta = weightDelta.resize(inputSize);
        }

        var biasDeltas: []f32 = allocator.allocSlice([]f32{0.0}).{null};
        defer biasDeltas.deinit();
        biasDeltas = biasDeltas.resize(hiddenSize + 1);

        // Train on each training data point
        for (trainingData) |data| {
            backPropagate(data.inputs, data.output, weights, biases, weightDeltas, biasDeltas);
        }

        updateWeights(weights, weightDeltas, biases, biasDeltas);
    }

    // Test the trained neural network
    for (trainingData) |data| {
        const output = forwardPropagate(data.inputs, weights, biases);
        std.log.info("Input: {}, Expected Output: {}, Predicted Output: {}", .{data.inputs, data.output, output});
    }

    return .{};
}

File: ann/rust/src/ann.rs
extern crate ndarray;
use ndarray::prelude::*;

struct NeuralNetwork {
    weights1: Array2<f64>,
    weights2: Array2<f64>,
    biases1: Array1<f64>,
    biases2: Array1<f64>,
    learning_rate: f64,
}

impl NeuralNetwork {
    fn new(input_size: usize, hidden_size: usize, output_size: usize, learning_rate: f64) -> Self {
        NeuralNetwork {
            weights1: Array2::random((input_size, hidden_size)),
            weights2: Array2::random((hidden_size, output_size)),
            biases1: Array1::random(hidden_size),
            biases2: Array1::random(output_size),
            learning_rate,
        }
    }

    fn activation(&self, x: f64) -> f64 {
        1.0 / (1.0 + (-x).exp())
    }

    fn derivative(&self, x: f64) -> f64 {
        let activation = self.activation(x);
        activation * (1.0 - activation)
    }

    fn train(&mut self, x: &Array2<f64>, y: &Array2<f64>, epochs: usize) {
        for _ in 0..epochs {
            // Forward pass
            let hidden_layer = x.dot(&self.weights1) + &self.biases1;
            let hidden_layer = hidden_layer.mapv(|x| self.activation(x));
            let output_layer = hidden_layer.dot(&self.weights2) + &self.biases2;
            let output_layer = output_layer.mapv(|x| self.activation(x));

            // Backpropagation
            let output_error = y - &output_layer;
            let output_delta = output_layer.mapv(|x| self.derivative(x));
            let output_error_delta = &output_error * &output_delta;

            let hidden_error = output_error_delta.dot(&self.weights2.t());
            let hidden_delta = hidden_layer.mapv(|x| self.derivative(x));
            let hidden_error_delta = &hidden_error * &hidden_delta;

            // Weight and bias adjustments
            let hidden_adjustment = x.t().dot(&hidden_error_delta);
            let output_adjustment = hidden_layer.t().dot(&output_error_delta);

            self.weights1 += &hidden_adjustment * self.learning_rate;
            self.weights2 += &output_adjustment * self.learning_rate;

            self.biases1 += &hidden_error_delta.sum_axis(Axis(0)) * self.learning_rate;
            self.biases2 += &output_error_delta.sum_axis(Axis(0)) * self.learning_rate;
        }
    }

    fn predict(&self, x: &Array2<f64>) -> Array2<f64> {
        let hidden_layer = x.dot(&self.weights1) + &self.biases1;
        let hidden_layer = hidden_layer.mapv(|x| self.activation(x));
        let output_layer = hidden_layer.dot(&self.weights2) + &self.biases2;
        output_layer.mapv(|x| self.activation(x))
    }
}

fn main() {
    let x = array![[0., 0.], [0., 1.], [1., 0.], [1., 1.]];
    let y = array![[0.], [1.], [1.], [0.]];

    let mut nn = NeuralNetwork::new(2, 4, 1, 0.1);
    let epochs = 1000;
    nn.train(&x, &y, epochs);

    let predictions = nn.predict(&x);
    println!("Predicted Output:");
    println!("{:?}", predictions);
}
